### PR TITLE
Validations

### PR DESCRIPTION
**Pull request by mudasiri**
In this PR, I did the following:

Added validations for models
Added units test for models

### CODE REVIEW

File: spec/models/comment_spec.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+require 'rails_helper'
+
+RSpec.describe Comment, type: :model do
+  let(:user) { User.create(name: 'Tom') }
+  let(:post) { Post.new(author: user, text: 'This is some text', comments_counter: 1) }
+  let(:comment) { Comment.new(post:, author: user, text: 'Hi Tom!') }
+
+  it 'increments the post comments counter' do
+    post.increment!(:comments_counter)
```
Comment (franclobo) [inline spec/models/comment_spec.rb:9]: - [ ] Please 🙏🏽 correct the migrations to be able to create the correct tests for comments validations 👇🏽 

```bash
== 20230609141944 AddPostRefToComment: migrating ==============================
-- add_reference(:comments, :post_id, {:null=>false, :foreign_key=>true})
rails aborted!
StandardError: An error has occurred, this and all later migrations canceled:

PG::UndefinedTable: ERROR:  relation "post_ids" does not exist
/home/franclobo/microverse/mybloggerapp/db/migrate/20230609141944_add_post_ref_to_comment.rb:5:in `change'

Caused by:
ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR:  relation "post_ids" does not exist
/home/franclobo/microverse/mybloggerapp/db/migrate/20230609141944_add_post_ref_to_comment.rb:5:in `change'

Caused by:
PG::UndefinedTable: ERROR:  relation "post_ids" does not exist
/home/franclobo/microverse/mybloggerapp/db/migrate/20230609141944_add_post_ref_to_comment.rb:5:in `change'
Tasks: TOP => db:migrate
```
The correct migration should be:
```ruby
# frozen_string_literal: true

class AddPostRefToComment < ActiveRecord::Migration[7.0]
  def change
    add_reference :comments, :posts, null: false, foreign_key: true
  end
end
```

```bash
== 20230609142126 AddUserRefToComment: migrating ==============================
-- add_reference(:comments, :author_id, {:null=>false, :foreign_key=>true})
rails aborted!
StandardError: An error has occurred, this and all later migrations canceled:

PG::UndefinedTable: ERROR:  relation "author_ids" does not exist
/home/franclobo/microverse/mybloggerapp/db/migrate/20230609142126_add_user_ref_to_comment.rb:5:in `change'

Caused by:
ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR:  relation "author_ids" does not exist
/home/franclobo/microverse/mybloggerapp/db/migrate/20230609142126_add_user_ref_to_comment.rb:5:in `change'

Caused by:
PG::UndefinedTable: ERROR:  relation "author_ids" does not exist
/home/franclobo/microverse/mybloggerapp/db/migrate/20230609142126_add_user_ref_to_comment.rb:5:in `change'
Tasks: TOP => db:migrate
```

The correct migration should be:

```
# frozen_string_literal: true

class AddUserRefToComment < ActiveRecord::Migration[7.0]
  def change
    add_reference :comments, :author, column: :author_id, null: false, foreign_key: { to_table: :users }
  end
end
```

File: spec/models/like_spec.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+require 'rails_helper'
+
+RSpec.describe Like, type: :model do
+  let(:user) { User.create(name: 'Tom') }
+  let(:post) { Post.create(author: user, text: 'This is my first post', likes_counter: 1) }
+
+  it 'increments the post likes counter' do
+    post.increment!(:likes_counter)
```
Comment (franclobo) [inline spec/models/like_spec.rb:8]: - [ ] Similar case with the comments, please 🙏🏽 correct the migrations to be able to create the correct validations tests.

```bash
== 20230609142239 AddUserRefToLike: migrating =================================
-- add_reference(:likes, :author_id, {:null=>false, :foreign_key=>true})
rails aborted!
StandardError: An error has occurred, this and all later migrations canceled:

PG::UndefinedTable: ERROR:  relation "author_ids" does not exist
/home/franclobo/microverse/mybloggerapp/db/migrate/20230609142239_add_user_ref_to_like.rb:5:in `change'

Caused by:
ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR:  relation "author_ids" does not exist
/home/franclobo/microverse/mybloggerapp/db/migrate/20230609142239_add_user_ref_to_like.rb:5:in `change'

Caused by:
PG::UndefinedTable: ERROR:  relation "author_ids" does not exist
/home/franclobo/microverse/mybloggerapp/db/migrate/20230609142239_add_user_ref_to_like.rb:5:in `change'
Tasks: TOP => db:migrate
```

```bash
== 20230609142450 AddPostRefToLike: migrating =================================
-- add_reference(:likes, :post_id, {:null=>false, :foreign_key=>true})
rails aborted!
StandardError: An error has occurred, this and all later migrations canceled:

PG::UndefinedTable: ERROR:  relation "post_ids" does not exist
/home/franclobo/microverse/mybloggerapp/db/migrate/20230609142450_add_post_ref_to_like.rb:5:in `change'

Caused by:
ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR:  relation "post_ids" does not exist
/home/franclobo/microverse/mybloggerapp/db/migrate/20230609142450_add_post_ref_to_like.rb:5:in `change'

Caused by:
PG::UndefinedTable: ERROR:  relation "post_ids" does not exist
/home/franclobo/microverse/mybloggerapp/db/migrate/20230609142450_add_post_ref_to_like.rb:5:in `change'
Tasks: TOP => db:migrate
```

File: spec/models/post_spec.rb
Comment (franclobo) [inline spec/models/post_spec.rb:1]: - [ ] Similar case with the comments and likes, please 🙏🏽 fix the migrations before creating the validations tests.

```bash
== 20230609142346 AddUserRefToPost: migrating =================================
-- add_reference(:posts, :author_id, {:null=>false, :foreign_key=>true})
rails aborted!
StandardError: An error has occurred, this and all later migrations canceled:

PG::UndefinedTable: ERROR:  relation "author_ids" does not exist
/home/franclobo/microverse/mybloggerapp/db/migrate/20230609142346_add_user_ref_to_post.rb:5:in `change'

Caused by:
ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR:  relation "author_ids" does not exist
/home/franclobo/microverse/mybloggerapp/db/migrate/20230609142346_add_user_ref_to_post.rb:5:in `change'

Caused by:
PG::UndefinedTable: ERROR:  relation "author_ids" does not exist
/home/franclobo/microverse/mybloggerapp/db/migrate/20230609142346_add_user_ref_to_post.rb:5:in `change'
Tasks: TOP => db:migrate
```

File: app/models/user.rb [hunk 3]
```diff
@@ -3,6 +3,9 @@ class User < ApplicationRecord
   has_many :likes, foreign_key: 'author_id'
   has_many :comments, foreign_key: 'author_id'
```
Comment (franclobo) [inline app/models/user.rb:4]: - [ ] To use the foreign_key: 'author_id' we should define the column author_id in the corresponding migration, like this 👇🏽 

```ruby
# frozen_string_literal: true

class AddUserRefToComment < ActiveRecord::Migration[7.0]
  def change
    add_reference :comments, :author, column: :author_id, null: false, foreign_key: { to_table: :users }
  end
end
```

File: app/models/post.rb [hunk 4]
```diff
@@ -5,6 +5,10 @@ class Post < ApplicationRecord
 
   after_create :increment_user_posts_counter
 
+  validates :title, presence: true, length: { maximum: 250 }
+  validates :comments_counter, numericality: { only_inter: true, greater_than_or_equal_to: 0 }
```
Comment (franclobo) [inline app/models/post.rb:9]: - [ ] To use comment**s**_counter we should define the column in the corresponding migration 👇🏽 

```ruby
# frozen_string_literal: true

class CreatePost < ActiveRecord::Migration[7.0]
  def change
    create_table :posts do |t|
      t.string :title
      t.string :text
      t.integer :comments_counter
      t.integer :likes_counter

      t.timestamps
    end
  end
end
```

**Review by franclobo — CHANGES_REQUESTED** [review]
Hi @mudasiri, 

## ♻️ Required Changes ♻️

istockphoto-1079018918-612x612

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights

- No linter errors ✔️ 
- Descriptive PR ✔️ 
- Correct Git flow ✔️ 
- Good validations ✔️ 


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/models/comment_spec.rb (hunk 1 above)
Comment (mudasiri) [reply spec/models/comment_spec.rb:9]: Updated the names

File: db/migrate/20230609142126_add_user_ref_to_comment.rb [hunk 5]
```diff
@@ -2,6 +2,6 @@
 
 class AddUserRefToComment < ActiveRecord::Migration[7.0]
   def change
-    add_reference :comments, :author_id, null: false, foreign_key: true
+    add_reference :comments, :author, column: :author_id, null: false, foreign_key: { to_table: :users }
```
Comment (andersonlebon) [inline db/migrate/20230609142126_add_user_ref_to_comment.rb:5]: image

- [ ] Kindly make sure solve all your migration errors the same way you did in the above. check in all the migration that are referencing the user model, the should specify the table. like this

for the post table
```rb
add_reference :posts, :author, column: :author_id, null: false, foreign_key: { to_table: :users }
```

for the comment  table
```rb
add_reference :comments, :author, column: :author_id, null: false, foreign_key: { to_table: :users }
```
for the like table
```rb
add_reference :likes, :author, column: :author_id, null: false, foreign_key: { to_table: :users }
```

**Review by andersonlebon — CHANGES_REQUESTED** [review]
## Hi @mudasiri 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

almost-there

## Required Changes ♻️
_Check the comments under the review._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question `@andersonlebon` so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: db/migrate/20230609140544_create_post.rb [hunk 6]
```diff
@@ -5,9 +5,9 @@ def change
     create_table :posts do |t|
       t.string :title
       t.string :text
-      t.integer :comment_counter
+      t.integer :comments_counter
       t.integer :likes_counter
-
+      add_reference :posts, :author, column: :author_id, null: false, foreign_key: { to_table: :users }
```
Comment (addod19) [inline db/migrate/20230609140544_create_post.rb]: - [ ] Kindly move this piece of code outside the `create_table :posts` block, let the code be between `line12 & line13`
Screenshot from 2023-06-10 11-38-29

File: db/migrate/20230609140917_create_comment.rb [hunk 7]
```diff
@@ -4,7 +4,7 @@ class CreateComment < ActiveRecord::Migration[7.0]
   def change
     create_table :comments do |t|
       t.string :text
-
+      add_reference :comments, :author, column: :author_id, null: false, foreign_key: { to_table: :users }
```
Comment (addod19) [inline db/migrate/20230609140917_create_comment.rb]: - [ ] Kindly move this piece of code outside the `create_table :comments` block, let the code be between `line12 & line13`
This is how our code should like like but not exactly the same
```
create_table :comments do |t|
      t.text :text
      t.integer :author_id
      t.integer :post_id

      t.timestamps
end
add_foreign_key :comments, :users, column: :author_id
add_foreign_key :comments, :posts, column: :post_id
add_index :comments, :author_id
add_index :comments, :post_id
```

**Review by addod19 — CHANGES_REQUESTED** [review]
Hi @mudasiri , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
- [ ] Nice job addressing some of the comments from the previous reviews :100: 


 ## Required Changes ♻️
_Check the comments under the review._
- [ ] Here is a [link](https://guides.rubyonrails.org/active_record_migrations.html) to help us understand more about active record migrations

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by mudasiri** [comment]
@addod19  please I have exhausted all my reviews, so kindly re-check my update and approve it here so I can submit it.

**Review by franclobo — APPROVED** [review]
Hi @mudasiri, 

# STATUS: APPROVED 😃 

git-merge

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

- No linter errors ✔️ 
- Correct Git flow ✔️ 
- Descriptive PR ✔️ 
- Correct migrations ✔️ 
- All tests passed 💯 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**